doc/_extensions/ncs_migration_guide_extraction.py

```diff
@@ -0,0 +1,490 @@
+                break
+            new_guides.append(git_show_file(rev_end, guide))
+
+        return new_guides
+
+    start_version_file = git_show_file(rev_start, "VERSION")
+    end_version_file = git_show_file(rev_end, "VERSION")
+    if not start_version_file or not end_version_file:
+        print("One of the version files could not be found")
+        return None
+
+    start_version = get_zephyr_version(start_version_file)
+    end_version = get_zephyr_version(end_version_file)
+
+    all_guides = {}
+    for guide in git("ls-tree", "--name-only", rev_end, "doc/releases/", cwd=cwd).split(
+        "\n"
+    ):
+        m = re.match(r"doc/releases/migration-guide-(\d+)\.(\d+)\.rst$", guide)
+        if m:
+            all_guides[tuple(map(int, m.groups()))] = guide
+
+    # The latest version of the previous migration guide
+    guide_file = get_guide_file(rev_end, start_version)
+
+    # Previous migration guide at the time of rev_start
+    prev_guide_file = get_guide_file(rev_start, start_version)
+
+    # Latest version of the latest migration guides
+    next_guide_files = get_new_migration_guides(start_version, end_version, all_guides)
+
+    return assemble_migration_guide(prev_guide_file, guide_file, next_guide_files)
+
+
+def assemble_migration_guide(
+    base: str, current: str, next: List[str]
+) -> Dict[str, Dict[str, List[str]]]:
+    """Collect changes made since the second-last downstream release to the last release.
+
+    The changes include everything in the `next` list, and the differences from
+    `base` to `current`.
+
+    The returned data structure is a dictionary mapping all required and recommended
+    changes to a mapping of titles to paragraphs that should belong to the guide
+    under that title. Paragraphs that do not belong to a title are listed under
+    'untitled'.
+
+    Args:
+        base: The latest migration guide at the start revision.
+        current: The base migration guide at the end revision.
+        next: List of new migration guides introduced since the start revision.
+
+    Returns:
+        All new paragraphs in a data structure representing the hierarchy
+        required|recommended -> title -> paragraph
+    """
+
+    def flatten(parsed_guide: Dict[str, Dict[str, List[str]]]) -> Dict[str, List[str]]:
+        flattened = {}
+        for change_type in ("required", "recommended"):
+            if not change_type in parsed_guide:
+                flattened[change_type] = []
+            else:
+                flattened[change_type] = [
+                    section
+                    for title_sections in parsed_guide[change_type].values()
+                    for section in title_sections
+                ]
+        return flattened
+
+    def add_difference(base, new, add_to_base=False) -> Dict[str, Dict[str, List[str]]]:
+        target = base if add_to_base else {}
+        flattened = flatten(base)
+
+        for change_type in ("required", "recommended"):
```

Comment:
Aren't we going to remove "required"/"recommended" sections upstream?

In general, I'm very skeptical about this change. We already build Zephyr docs, and all the migration guides will already be there. If the problem is that we never pick a released Zephyr... well... let's just discuss whether this makes sense in a first place instead of building more tooling on top.